presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/OrcFileWriter.java

```diff
@@ -0,0 +1,144 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class OrcFileWriter
+        implements Closeable
+{
+    private final PageBuilder pageBuilder;
+    private final OrcWriter orcWriter;
+
+    private boolean closed;
+    private long rowCount;
+    private long uncompressedSize;
+
+    public OrcFileWriter(
+            long chunkId,
+            List<Long> columnIds,
+            List<Type> columnTypes,
+            CompressionType compressionType,
+            File target,
+            TypeManager typeManager,
+            OrcWriterStats stats)
+    {
+        StorageTypeConverter converter = new StorageTypeConverter(typeManager);
+        List<Type> storageTypes = columnTypes.stream()
+                .map(converter::toStorageType)
+                .collect(toImmutableList());
+
+        this.pageBuilder = new PageBuilder(storageTypes);
+
+        Map<String, String> metadata = OrcFileMetadata.from(chunkId, columnIds, columnTypes).toMap();
+        this.orcWriter = createOrcFileWriter(target, columnIds, storageTypes, compressionType, metadata, stats);
+    }
+
+    public void appendPages(List<Page> pages)
+    {
+        for (Page page : pages) {
+            appendPage(page);
+        }
+    }
+
+    public void appendPages(List<Page> pages, int[] pageIndexes, int[] positionIndexes)
+    {
+        checkArgument(pageIndexes.length == positionIndexes.length, "pageIndexes and positionIndexes do not match");
+
+        for (int i = 0; i < pageIndexes.length; i++) {
+            Page page = pages.get(pageIndexes[i]);
+            int position = positionIndexes[i];
+            appendPositionTo(page, position, pageBuilder);
+
+            if (pageBuilder.isFull()) {
+                appendPage(pageBuilder.build());
+                pageBuilder.reset();
+            }
+        }
+
+        if (!pageBuilder.isEmpty()) {
+            appendPage(pageBuilder.build());
+            pageBuilder.reset();
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        if (closed) {
+            return;
+        }
+        closed = true;
+
+        try {
+            orcWriter.close();
+        }
+        catch (IOException e) {
+            throw new PrestoException(RAPTOR_STORAGE_ERROR, "Failed to close writer", e);
+        }
+    }
+
+    public long getRowCount()
+    {
+        return rowCount;
+    }
+
+    public long getUncompressedSize()
+    {
+        return uncompressedSize;
+    }
+
+    private void appendPage(Page page)
+    {
+        rowCount += page.getPositionCount();
+        uncompressedSize += page.getLogicalSizeInBytes();
+
+        try {
+            orcWriter.write(page);
+        }
+        catch (IOException e) {
+            throw new PrestoException(RAPTOR_STORAGE_ERROR, "Failed to write data", e);
+        }
+    }
+
+    private static void appendPositionTo(Page page, int position, PageBuilder pageBuilder)
```

Comment:
I think ORC writer has its own buffer so we don't need to flush until we reach a full page. Directly call `Writer::write` shouldn't harm